<a href="https://colab.research.google.com/github/cjaucian/MAT421/blob/main/ModuleA_9_1_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import matplotlib.pyplot as plt


**9.1 Base-N and Binary**

Numbers can be expressed in different base systems, we commonly use base-10 which indicates that each individual number is a factor or 10^n. For example:
101(base10) = 1*10^2 + 0*10^1 + 1*10^0

We can convert any number in base-10 to another base. For example, if we wanted to write 12 in base 3, we are looking to use only 0, 1, 2.
12(base 10) = 9 + 3 + 0 = 1*3^2 + 1*3^1 + 0*3^0 = 110(base 3)

Converting numbers to binary, which is most commonly used in computers means that we use only 0, and 1. For example:
31(base 10) = 16 + 8 + 4 + 2 + 1 = 1*2^4 + 1*2^3 + 1*2^2 + 1*2^1 + 1*2^0 = 11111(base 2)

Converting numbers from base-10 to binary is time consuming and tedious. But we can use Python to make the process easier. We can even computer the products, sums, differences and quotients of binary numbers in Python. Similarly to the way we would compute them in base-10.

Below are examples of converting numbers in Base-N and Binary, as well as operations in both representations.

**Example 1:** Converting numbers in base-10 to binary.

In [ ]:
print(1211,'=',bin(1211),'\n',12,'=',bin(12),'\n',552,'=',bin(552))


1211 = 0b10010111011 
 12 = 0b1100 
 552 = 0b1000101000


1211(base 10) is 10010111011 in binary.
 12(base 10) is 1100 in binary.
 552(base 10) is 1000101000 in binary.



 Binary numbers can be used in mathematical operations, similar to base-10 numbers.

 **Example 2:** Operations with Binary Numbers

In [ ]:
x=0b10010111011
y=0b1100
z=0b1000101000

print('0b10010111011 * 0b1100 =',x*y)
print('0b1100 / 0b1000101000 =',y/z)
print('0b1100 + 0b1000101000 =',y+z)
print('0b10010111011 - 0b1000101000 =',x-z)

0b10010111011 * 0b1100 = 14532
0b1100 / 0b1000101000 = 0.021739130434782608
0b1100 + 0b1000101000 = 564
0b10010111011 - 0b1000101000 = 659


0b10010111011 * 0b1100 = 14532

0b1100 / 0b1000101000 = 0.021739130434782608

0b1100 + 0b1000101000 = 564

0b10010111011 - 0b1000101000 = 659

**9.2 Floating Point Numbers**

Binary representations do not give sufficient accuracy in engineering calculations. For this reason, we need to use a new number representation to give greater accuracy but remain within 64-bits. This representation is called floating point numbers.

Unlike binary numbers which represent each coefficient as a power of 2, floats use three different parts. These three parts are: sign indicator, exponent, and fraction. The sign indicator, s, show whether a number is positive or negative. The exponent, e, for a base of 2. Lastly, the fraction, f, is the coefficient of the exponent.

n = (-1)^2 * 2^(e-1023)*(1+f)

In [1]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

The spacing function in Python calculates the distance between one number to the next, known as the gap. The fraction in the float is being multiplied by 2^(e-1023), which creates a larger and larger gap as the number grows larger.

**Example** Using the spacing function we will calculate the gap at 1e9.

In [3]:
import numpy as np
np.spacing(1e9)

1.1920928955078125e-07

1.1920928955078125e-07

If we add a number that is less than one-half the size of the gap at 1e9, Python will still show the same value.

In [5]:
import numpy as np
1e9 == (1e9 + np.spacing(1e9)/3)

True

True

There are limits to the values that can be represented by the floating point numbers. A number that exceeds the maximum float value will show as infinity, and is considered an overflow. A number that is smaller than the smallest subnormal number will be assigned 0, and is considered underflow.

In [6]:
sys.float_info.max

1.7976931348623157e+308

1.7976931348623157e+308

In [7]:
sys.float_info.min

2.2250738585072014e-308

2.2250738585072014e-308

**Example** Show that a subnormal number is represented by 0.0.

In [8]:
2**(-2000)

0.0

0.0

**Example** Show that a subnormal number cannot be distinguished from 0.0.

In [9]:
2**(-2000) == 0

True

**9.3 Round-off Errors**

In section 9.2, it was stated that computers represent floating point numbers as base 2 fractions. This representation creates errors because the number is not recorded with perfect precision. The difference between the approximation of a number and the correct value of a number is called round-off error.

Round-off errors occur when we represent infinite floating point numbers, like pi or 1/9. Since we have to use a finite number of digits, we end up with a value that could be more or less than the correct value.

In Python, we will get a different calculation than what we expect in arithmetic computations. If we take 3.9 - 3.8, we would expect to get 0.1.

In [11]:
3.9 - 3.8 == 0.1

False

False

So if 3.9 - 3.8 does not equal 0.1, what does it equal?

In [12]:
3.9 - 3.8

0.10000000000000009

0.10000000000000009

**Example** Show that 0.4 + 0.3 + 0.2 does not equal 0.9.

In [16]:
0.4 + 0.3 + 0.2 == 0.9

False

False

We can use the round function to get inexact values to be comparable to another.

**Example** Using the round function to get values to compare.

In [17]:
round(0.4 + 0.3 + 0.2, 5) == round(0.9, 5)

True

True

In arithmetic, we can accumulate round-off errors since we round each calculation to a finite number of values.

**Example** Computation value only taken once.

In [26]:
1 + 1/9 - 1/9

1.0

If we complete an addition and subtraction problem more than once, 100 times, or even 1000 times. The round-off error will accumulate and grow larger.

In [29]:
def add_and_subtract(iterations):
  result = 1

  for i in range(iterations):
    result += 1/9

  for i in range(iterations):
    result -= 1/9
  return result


print(add_and_subtract(100))
print(add_and_subtract(1000))
print(add_and_subtract(10000))

1.0000000000000004
1.0000000000000022
1.0000000000000981


1.0000000000000004

1.0000000000000022

1.0000000000000981